modules/aggregations/src/main/java/org/elasticsearch/aggregations/bucket/timeseries/TimeSeriesAggregator.java

```diff
@@ -92,15 +92,29 @@ protected void doClose() {
     protected LeafBucketCollector getLeafCollector(AggregationExecutionContext aggCtx, LeafBucketCollector sub) throws IOException {
         return new LeafBucketCollectorBase(sub, null) {
 
+            // This helps significantly reducing time spent attempting to add bucket + tsid combos that already were added.
+            long currentTsidOrd = -1;
+            long currentBucket = -1;
+            long currentBucketOrdinal;
+
             @Override
             public void collect(int doc, long bucket) throws IOException {
+                if (currentBucket == bucket && currentTsidOrd == aggCtx.getTsidOrd()) {
+                    collectExistingBucket(sub, doc, currentBucketOrdinal);
+                    return;
+                }
+
                 long bucketOrdinal = bucketOrds.add(bucket, aggCtx.getTsid());
```

Comment:
If we have a tsidOrd can we use that as a key instead of the bytes?

Reply:
So we need the tsid as bytes in the `buildAggregations()` method and I don't think we can use this tsidord that `TimeSeriesIndexSearcher` generates as a lookup to get the tsid?

Reply:
Indeed this tsidord is neither a segment ordinal nor a global ordinal but an ordinal for TSIDs that intersect with the query on this shard. So it can't be used as a key to retrieve TSIDs from the terms dictionary. (Separately, maybe we should rename this method to reduce chances that someone mistakenly uses it as an ordinal for the `_tsid` terms dict? I worry a bit that it would look like it works for common cases like the query only consists of a range query on the `@timestamp` and this time range happens to match every unique TSID, so tests might not even catch the problem if someone started doing this?)

Reply:
Let's rename the method in a follow up change?